### ISSUE TITLE
Efficiently store and search for emojis

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#12189]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Start typing `:smile`

## Expected Result:
The soon-to-be-implemented emoji suggestions UI should produce not only results that _begin_ with `smile`, like `smile` and `smiley`, but also things that contain `smile` as a substring, such as `sweat_smile`.

## Actual Result:
Only emoji that _begin_ with smile will come up in the search results.

## Workaround:
n/a

## Platform:
All

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b83383f908c1ef35</li>
        <li>Upwork Job ID: 1613221194285375488</li>
        <li>Last Price Increase: 2023-01-11</li>
    </ul>
</details>

**Comment by roryabraham** [comment]
cc @stitesExpensify in case you want to be co-assigned on this one as well

**Comment by roryabraham** [comment]
This is hinted at in the issue title, but I think that a [suffix tree](https://www.educative.io/answers/what-is-a-suffix-tree) will be the appropriate data structure to use for this search, since the emoji dataset is so large (and will likely continue to grow over time)

**Comment by roryabraham** [comment]
Also, note that this does not necessarily have to be on HOLD for https://github.com/Expensify/App/issues/12188, because even though we don't have the UI yet, we should be able to verify the correct behavior with unit tests, which will be a _hard requirement_ for any PR to implement this issue.

**Comment by roryabraham** [comment]
@adelekennedy I think $500 is an appropriate starting value for this issue

**Comment by leeoniya** [comment]
> there are a few really good and fast fuzzy search libs out there. One that comes to mind is fuse.js

i wouldn't call Fuse fast, exactly. but i've done a pretty broad comparison of fuzzy match libs in the course of building uFuzzy.

https://github.com/leeoniya/uFuzzy#performance

**Comment by stitesExpensify** [comment]
I agree with @rushatgabhane about not using a fuzzy search. For reference the Trie which is currently implemented finds emojis in [1-6ms](https://github.com/Expensify/App/pull/10165#discussion_r976521139). 

I think a suffix tree as mentioned in the OP (which is very similar to a trie) is the correct data structure, and we should stick to that.

**Comment by leeoniya** [comment]
:+1: excited to see what you folks come up with! i'd be interested to add a Ukkonen suffix tree implementation to the bench table as well. btw, uFuzzy is not based on string distance metrics, since that typically produces garbage results!

out of curiosity i loaded up all the emoji keywords from https://github.com/Expensify/App/blob/main/assets/emojis.js and got 0.3ms matches, or 0.7ms for out-of-order partial terms. this is with 0 mem overhead, since there's no index.

![image](https://user-images.githubusercontent.com/43234/199123496-1b928535-2f72-4ed2-b9a6-ca94dcfa44df.png)

**Comment by stitesExpensify** [comment]
@leeoniya those results are crazy fast! If you want to create a proposal, that sounds very promising!

I am a bit confused how there is 0 memory overhead since we would still have to create the data structure right? I'm also curious how long it takes to create that structure (seems like it should be fast using Ukkonen's algo).

**Comment by leeoniya** [comment]
the data structure is just the haystack string[]. if you dont have this array in mem already and have to concat/extract the keywords arrays, there would be some minimal additional allocs up front, but uFuzzy itself doesnt do anything to your haystack once it's made. the lib is regexp based and just processes that haystack in a linear/dumb fashion. pretty wild how fast it turned out! 🤯

**Comment by Karim-30** [comment]
@roryabraham we already have this feature because we are inserting the emoji name and its keywords into the [Trie](https://github.com/Expensify/App/blob/72fa14d878281b3db1a1fa28330b9fba835a7436/src/libs/EmojiTrie.js#L8-L32)

i.e when inserting this emoji 
```javascript
    {
        code: '😀',
        keywords: [
            'face',
            'grin',
        ],
        name: 'grinning',
    },
```
we aren't inserting a single emoji `{name:'grinning', {code: '😀'}}`

actually we are inserting something like this 
```javascript
{name:'grinning', {code: '😀'}}
{name:'face', {code: '', suggestions: [name:'grinning', {code: '😀'}]}
{name:'grin', {code: '', suggestions: [name:'grinning', {code: '😀'}]}
```

so when typing `:fa` in the composer we will see the `grinning` emoji

https://user-images.githubusercontent.com/108357004/199466664-d8475d43-95f9-4351-9fc4-389b739dfd9b.mov

**Comment by JmillsExpensify** [comment]
@adelekennedy I'm going to snag this one from you if you don't mind, as I'd like to help with the broader tracking issue.

**Comment by leeoniya** [comment]
i grabbed your current Trie implementation to bench it against single terms: https://github.com/leeoniya/uFuzzy/tree/emojis-trie

i think @stitesExpensify's statement that your current code takes 1-6ms was likely measuring more than just the core search algorithm.

- the trie construction takes about 20ms (vs 0.5ms to boot uFuzzy)
- the trie retained ram use is ~2.9mb (vs 1.9mb for uFuzzy)
- the trie bench time is ~30ms (vs 110ms for uFuzzy)

so the trie is definitely faster than a regex match, at the cost of more mem and more index build time, as expected! when the overall numbers for a single match are this low in both cases (sub 1ms), the perf difference can probably be disregarded. no one will notice a 20ms boot time either. this may change significantly for much larger datasets since the trie size and init will grow more than linearly with dataset size.

the big difference of course is match quality. the trie only handles exact prefix matches and cannot handle multiple/combined partial terms, out of order matching, with minor misspellings. for that you need something more sophisticated. maybe that's not needed for this case, and that's a perfectly valid choice as well :+1: 

you can run the bench locally on that branch via these urls after running `http-server` (from npm) in the repo root dir:

http://localhost:8080/demos/compare.html?libs=uFuzzy&lists=emojis&bench
http://localhost:8080/demos/compare.html?libs=EmojisTrie&haystack=emojis&bench

**Comment by stitesExpensify** [comment]
@leeoniya Am I correct that a fuzzy search will get slower with larger datasets (which will certainly happen when we allow users to add custom emojis), whereas a suffix tree would not see any impact due to the search time being linear to the size of the _search string_ rather than the size of the dataset? 

I also wanted to mention that the suffix tree solution we recommended in the OP is both faster and takes less space than the trie we currently have implemented.

**Comment by cead22** [comment]
I think fuzzy search makes sense, so typing things with small typos still gives you the match you're looking for, like @stitesExpensify last example

**Comment by roryabraham** [comment]
I'm less concerned about initialization time than search time, because that happens only once on app initialization, and our current benchmarks for that are quite fast (~50ms). If if ever became a problem, we could also lazy-load the EmojiTrie so as not to block any UI except the emoji selector on its initialization.

**Comment by leeoniya** [comment]
> @leeoniya Am I correct that a fuzzy search will get slower with larger datasets (which will certainly happen when we allow users to add custom emojis), whereas a suffix tree would not see any impact due to the search time being linear to the size of the search string rather than the size of the dataset?

yeah, i would expect a trie of suffix tree to have better runtime scaling than fuzzy search, though it depends on how the fuzzy search is implemented (e.g. levenshtein in JS will scale worse than a regex that delegates to a C or asm backend).

> If we're deciding whether or not to use fuzzy search, we should give counter arguments to https://github.com/Expensify/App/issues/12189#issuecomment-1296272527 first.

tl;dr; `fuzzy !== levenshtein`. relying on distance metrics alone for fuzzy search is guaranteed to produce poor results. there are many better ways to do it.

> I'm less concerned about initialization time than search time, because that happens only once, on app initialization, and our current benchmarks for that are quite fast (~50ms). If if ever became a problem, we could also lazy-load the EmojiTrie so as not to block any UI except the emoji selector on its initialization.

tried with an emoji count of ~20k and get ~3.4ms per typed char with out-of-order partial terms. but yes, a couple passes over a suffix tree would be faster than this. maybe in the range of 1-2ms?. not sure how much that matters in practice vs the other trade-offs.

![image](https://user-images.githubusercontent.com/43234/199804713-6d1fb8b5-7585-4ed1-a4b8-a7cd793085b6.png)

**Comment by JmillsExpensify** [comment]
Still working through the ideal solution. @stitesExpensify for clarity, I'd imagine this is still not a priority until we close out the LHN API refactors?

**Comment by stitesExpensify** [comment]
That's correct. I'll get back to this ASAP

**Comment by roryabraham** [comment]
Was chatting a bit with @JmillsExpensify and I think it might be a good idea to put this on HOLD so we can have a broader discussion about search features. I know emojis are the case-study here, but it's probably wise to take a step back here, evaluate our current and future needs, then design a performant, hopefully future-proof solution for search that will scale to meet our needs beyond just emojis.

**Comment by JmillsExpensify** [comment]
@stitesExpensify This is probably perfect timing for you, now that the LHN API refactor is wrapped up. There isn't any immediate urgency for emoji search, so it'd be good to make sure where ever we land approximates the use cases we'll need across chats, requests, card transactions, attachments and similar biz/consumer use cases.

**Comment by mountiny** [comment]
I am not sure if the search logic we decide on here can influence much the search within the whole. We plan to use sqlite features/ sqlite search for that once logrock proves the concept 🤔  unless I am misunderstanding something here.

**Comment by roryabraham** [comment]
There are a lot of different ideas floating around in here, probably too many to the point where we need to chat about this in slack and come to consensus, and for that I think we need an internal champion. While I'm definitely interested, I don't have the bandwidth this week and @stitesExpensify is out of office, so I'm going to briefly put this on HOLD and make it a weekly. It's not super urgent so better to make sure we take our time and get it right.

Thanks for your input so far everyone so far and thanks for your patience.

**Comment by JmillsExpensify** [comment]
Coming back to this, I think that before we pick this back up, we need to get a better handle on the emoji-related bugs in the linked tracking issue – not to mention make significant progress in WAQ overall. I'm adding the monthly label because I expect it to take that long to get both in a more stable spot.

**Comment by JmillsExpensify** [comment]
Still on hold per the comment above.

**Comment by JmillsExpensify** [comment]
Still on hold based on our more pressing focus on bugs atm.

**Comment by JmillsExpensify** [comment]
WAQ should be wrapping up in the next week or so, and when it does, this is fair game again.

**Comment by JmillsExpensify** [comment]
Officially removing the HOLD. This would be an ideal one to consider picking up.

**Comment by JmillsExpensify** [comment]
We start all issues at $1,000 now, so I'd actually default to that instead. 😄 That said, @stitesExpensify now that this is off hold, what's our plan for this issue. Are we keeping it external and opening it up for proposals?

**Comment by JmillsExpensify** [comment]
Totally agree we should go the design doc route. For that reason, I'm going to remove the `external` and add `internal` for the time being.

**Comment by stitesExpensify** [comment]
Yes, this is not currently being discussed as we're focused on the emoji reactions doc. I'll move it back to weekly once it becomes a priority again

**Comment by JmillsExpensify** [comment]
Still lower in priority compared to emoji reactions.

**Comment by stitesExpensify** [comment]
Our colon code results could be improved quite a bit. Specifically that codes like `:smile` don't show `:sweat_smile`

**Comment by stitesExpensify** [comment]
Still waiting on this for now since it's just a polish issue

**Comment by JmillsExpensify** [comment]
For what it's worth, I have been hitting this one more and more. I agree it's a polish issue, it's probably just very easy to get used to emoji search on other apps, remember those search terms, and then come up short using our search. I'd like to see us re-prioritize this one. Could we open it up?

**Comment by JmillsExpensify** [comment]
Still not a priority.

**Comment by JmillsExpensify** [comment]
Same

**Comment by roryabraham** [comment]
What do we think here? Should we have an agency pick this one up?

**Comment by roryabraham** [comment]
Said differently, I think we should either move forward with this or just close it

**Comment by stitesExpensify** [comment]
Chatting with Rory 1:1, I think we should just close this. We can come back to it when we are doing custom emojis which are going to be stored on the server and have a lot of other considerations for search etc, so doing this now would just be doubling our work. Closing!
